### PR TITLE
[3/4] CUDNNv8 ResNet Fusion: Add fused_donv_drelu_dbn OP

### PR DESCRIPTION
**Pull request by Tom-Zheng**
### PR types

New features
### PR changes

OPs
### Description

This PR adds fused_dconv_drelu_dbn op, which is part of the CUDNNv8 ResNet fusion integration.

### CODE REVIEW

**Comment by Tom-Zheng** [comment]
@zyfncg Hi, would you please take a look?

**Comment by Tom-Zheng** [comment]
@zyfncg The CI failures will be approved manually and should not block review process. Please take a look.

**Comment by onecatcn** [comment]
2023-11-23 11:07:36 ****************
2023-11-23 11:07:36 0. You must have raindrops2sea or XiaoguangHu01 approval for change 20+ files or add than 1000+ lines of content.
2023-11-23 11:07:36 1. You must have one RD (XiaoguangHu01,chenwhql,zhiqiu,Xreki,luotao1,qili93,Aurelius84) approval for the usage of const_cast.
2023-11-23 11:07:36 2. paddle::get is not recommended for direct use, because it may throw an bad_variant_access exception without any stack information, so please use PADDLE_GET(_**)(dtype, value) series macros here. If these macros cannot meet your needs, please use try-catch to handle paddle::get and request chenwhql (Recommend), luotao1 or lanxianghit or Aurelius84 review and approve.
2023-11-23 11:07:36 3. Unittest is not allowed to be disabled.
2023-11-23 11:07:36 You must have one RD (kolinwei(Recommend), wanghuancoder, luotao1, QingshuChen, qili93 or ZzSean or Aurelius84) approval for the usage of @unittest.skip or @unittest.skipIf.
2023-11-23 11:07:36 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-11-23 11:07:36 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-11-23 11:07:36 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-11-23 11:07:36 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-11-23 11:07:36 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-11-23 11:07:36 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-11-23 11:07:36 4. It is an Op accuracy problem, please take care of it. You must have one RD (zhangting2020 (Recommend), luotao1 or phlrain, qili93, QingshuChen or Aurelius84) approval for the usage (either add or delete) of @skip_check_grad_ci. For more information, please refer to: https://github.com/PaddlePaddle/Paddle/wiki/Gradient-Check-Is-Required-for-Op-Test. The corresponding lines are as follows:

2023-11-24 11:27:18 [tools/ci_op_benchmark.sh:280] [INFO] Or you can apply for one RD (ZzSean(Recommend), JamesLim-sy, Xreki, luotao1) approval to pass this PR.

2023-11-23 12:51:03 Please find RD for approval first, and then find TPM for approval.
2023-11-23 12:51:03 0. You must have one RD (inference[ vivienfanghuagood(Recommend), yuanlehome, qingqing01 ] or slim[ wanghaoshuang(Recommend), qingqing01 ] or train[ chenwhql(Recommend), phlrain ]) approval for the changes of Inputs/Output/Attrs of OPs. The changes of OPs will cause that the new version inference fails to load model trained by the old version. Please modify your code. 
2023-11-23 12:51:03  For more details, please click [https://github.com/PaddlePaddle/Paddle/wiki/OP-Input-Output-Attribute-Compatibility-Modification].

**Review by zhangting2020 — DISMISSED** [review]

File: paddle/phi/infermeta/fusion.cc [hunk 1]
```diff
@@ -2041,7 +2041,139 @@ void FusedScaleBiasAddReluInferMeta(const MetaTensor& x1,
...
+                                     !!bn2_mean,
+                                     !!bn2_inv_std,
+                                     !!bn2_gamma,
+                                     !!bn2_beta,
+                                     !!bn2_input));
+  }
+  grad_weight->set_dims(weight.dims());
+  grad_bn1_input->set_dims(bn1_input.dims());
+  grad_bn1_gamma->set_dims(bn1_gamma.dims());
+  grad_bn1_beta->set_dims(bn1_beta.dims());
```
Comment (zyfncg) [inline paddle/phi/infermeta/fusion.cc]: please set `dtype` like this
image

**Review by wanghuancoder — DISMISSED** [review]
LGTM for  @unittest.skip or @unittest.skipIf

File: paddle/phi/kernels/fusion/gpu/fused_dconv_drelu_dbn_kernel.cu [hunk 2]
```diff
@@ -0,0 +1,1162 @@
+/* Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
Comment (zyfncg) [inline paddle/phi/kernels/fusion/gpu/fused_dconv_drelu_dbn_kernel.cu]: 2022 -> 2023

**Review by vivienfanghuagood — APPROVED** [review]
LGTM for add API

File: test/legacy_test/test_fused_dconv_drelu_dbn_op.py [hunk 3]
```diff
@@ -0,0 +1,478 @@
+#   Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
Comment (zyfncg) [inline test/legacy_test/test_fused_dconv_drelu_dbn_op.py]: 2022 ->2023

**Review by zyfncg — DISMISSED** [review]

File: test/legacy_test/test_fused_dconv_drelu_dbn_op.py (hunk 3 above)
Comment (Tom-Zheng) [reply test/legacy_test/test_fused_dconv_drelu_dbn_op.py]: done

File: paddle/phi/kernels/fusion/gpu/fused_dconv_drelu_dbn_kernel.cu (hunk 2 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/fusion/gpu/fused_dconv_drelu_dbn_kernel.cu]: done

File: paddle/phi/infermeta/fusion.cc (hunk 1 above)
Comment (Tom-Zheng) [reply paddle/phi/infermeta/fusion.cc]: done

**Review by zyfncg — APPROVED** [review]

**Comment by Tom-Zheng** [comment]
```
2023-11-30 11:48:08 0. You must have raindrops2sea or XiaoguangHu01 approval for change 20+ files or add than 1000+ lines of content.
2023-11-30 11:48:08 1. You must have one RD (XiaoguangHu01,chenwhql,zhiqiu,Xreki,luotao1,qili93,Aurelius84) approval for the usage of const_cast.
2023-11-30 11:48:08 2. paddle::get is not recommended for direct use, because it may throw an bad_variant_access exception without any stack information, so please use PADDLE_GET(_**)(dtype, value) series macros here. If these macros cannot meet your needs, please use try-catch to handle paddle::get and request chenwhql (Recommend), luotao1 or lanxianghit or Aurelius84 review and approve.
2023-11-30 11:48:08 3. Unittest is not allowed to be disabled.
2023-11-30 11:48:08 You must have one RD (kolinwei(Recommend), wanghuancoder, luotao1, QingshuChen, qili93 or ZzSean or Aurelius84) approval for the usage of @unittest.skip or @unittest.skipIf.
2023-11-30 11:48:08 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-11-30 11:48:08 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-11-30 11:48:08 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-11-30 11:48:08 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-11-30 11:48:08 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-11-30 11:48:08 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-11-30 11:48:08 4. It is an Op accuracy problem, please take care of it. You must have one RD (zhangting2020 (Recommend), luotao1 or phlrain, qili93, QingshuChen or Aurelius84) approval for the usage (either add or delete) of @skip_check_grad_ci. For more information, please refer to: https://github.com/PaddlePaddle/Paddle/wiki/Gradient-Check-Is-Required-for-Op-Test. The corresponding lines are as follows:
2023-11-30 11:48:08 test/legacy_test/test_fused_dconv_drelu_dbn_op.py
2023-11-30 11:48:08 +	@skip_check_grad_ci(reason="no grap op")
2023-11-30 11:48:08 +	@skip_check_grad_ci(reason="no grap op")
2023-11-30 11:48:08 +	@skip_check_grad_ci(reason="no grap op")
2023-11-30 11:48:08 +	@skip_check_grad_ci(reason="no grap op")
2023-11-30 11:48:08 +	@skip_check_grad_ci(reason="no grap op")
2023-11-30 11:48:08 +	@skip_check_grad_ci(reason="no grap op")
2023-11-30 11:48:08 There are 5 approved errors.
```

@XiaoguangHu01 @chenwhql @ZzSean @zhangting2020 Would you please approve?

**Comment by Tom-Zheng** [comment]
@qingqing01 Would you please approve for  PR-CI-Static-Check?

**Review by zhangting2020 — APPROVED** [review]

**Review by wanghuancoder — APPROVED** [review]

**Review by chenwhql — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Comment by Tom-Zheng** [comment]
@tianshuo78520a Would you please help approve `PR-CI-Coverage`

**Comment by Tom-Zheng** [comment]
@qingqing01 @heavengate  would you please approve for `PR-CI-Static-Check`?

**Review by heavengate — APPROVED** [review]

**Comment by Tom-Zheng** [comment]
@zyfncg Hi, would you please merge it?

**Pull request merged**